zerver/lib/retention.py

```diff
@@ -174,42 +163,28 @@ def move_attachments_message_rows_to_archive(msg_ids: List[int]) -> None:
         cursor.execute(query.format(message_ids=ids_list_to_sql_query_format(msg_ids)))
 
 def delete_messages(msg_ids: List[int]) -> None:
-    Message.objects.filter(
-        usermessage__isnull=True, id__in=msg_ids
-    ).delete()
-
-
-def delete_user_messages(usermsg_ids: List[int]) -> None:
-    UserMessage.objects.filter(id__in=usermsg_ids).delete()
-
+    Message.objects.filter(id__in=msg_ids).delete()
 
 def delete_expired_attachments() -> None:
     Attachment.objects.filter(
         messages__isnull=True, id__in=ArchivedAttachment.objects.all(),
     ).delete()
 
-def move_related_objects_to_archive(msg_ids: List[int]) -> List[int]:
-    usermsg_ids = move_user_messages_to_archive(msg_ids)
+def move_related_objects_to_archive(msg_ids: List[int]) -> None:
     move_models_with_message_key_to_archive(msg_ids)
     move_attachments_to_archive(msg_ids)
     move_attachments_message_rows_to_archive(msg_ids)
 
-    return usermsg_ids
-
-def clean_expired(msg_ids: List[int], usermsg_ids: List[int]) -> None:
-    delete_user_messages(usermsg_ids)
-    delete_messages(msg_ids)
-
 def archive_messages_by_recipient(recipient: Recipient, message_retention_days: int) -> None:
     msg_ids = move_expired_messages_to_archive_by_recipient(recipient, message_retention_days)
-    usermsg_ids = move_related_objects_to_archive(msg_ids)
-    clean_expired(msg_ids, usermsg_ids)
+    move_related_objects_to_archive(msg_ids)
+    delete_messages(msg_ids)
 
 def archive_personal_and_huddle_messages() -> None:
     for realm in Realm.objects.filter(message_retention_days__isnull=False):
         msg_ids = move_expired_personal_and_huddle_messages_to_archive(realm)
-        usermsg_ids = move_related_objects_to_archive(msg_ids)
-        clean_expired(msg_ids, usermsg_ids)
+        move_related_objects_to_archive(msg_ids)
+        delete_messages(msg_ids)
 
 def archive_stream_messages() -> None:
     streams = Stream.objects.filter(message_retention_days__isnull=False)
```

Comment:
If I read this code correctly, if one has an organization-level message retention policy, we don't do anything with it for streams, right?  

I feel like the correct behavior with our current options data model (which we may keep) is probably to use the organization-level retention policy for any streams that don't have an override to that policy.  Though I guess we don't have a way to express that -- `null` on a given stream could mean "forever retention policy for this stream" or "use default for organization".  I guess we probably want null to mean "use default for organization" and to use a special value like "-1" for "override and keep messages forever".  

I'm fine with merging this PR without resolving that, just wanted to make sure we're on the same page for where we're going.

Reply:
I was thinking we would set the retention policy for streams to the realm's policy (unless something else is set explicitly for the stream) by whatever mechanism we will have for setting realm/stream policies. But your described approach seems much better, so I added that behavior and a test for it in a new commit.